## Basics

1. Swap two numbers without using a third variable.

In [ ]:
a = 10
b = 20
print ("a:",a,"\nb:",b, end= " ")
a , b = b , a
print ("\nafter swapping")
print ("a:",a,"\nb:",b, end= " ")

a: 10 
b: 20 
after swapping
a: 20 
b: 10 

2. Take user's name and age as input and print them in a formatted sentence

In [ ]:
age = int(input("Enter Age:"))
name =input("Enter Name:")
print ("My name is ",name, "and my age is ", age)

Enter Age:10
Enter Name:ali
My name is  ali and my age is  10


3. Convert a float to an integer and print its data type.

In [ ]:
x = 1.000
y = int(x)
print(y)
print(type(y))

1
<class 'int'>


4. Take multiple inputs in one line and print them separately.

In [ ]:
a,b,c = input("Enter three values:").split()
print(a)
print(b)
print(c)


Enter three values:11 22 33
11
22
33


5. Check and print the data types of various inputs.

In [ ]:
a = input("Enter a value:")
print(type(a))
b = int(input("Enter a value:"))
print(type(b))
c = float(input("Enter a value:"))
print(type(c))

Enter a value:123
<class 'str'>
Enter a value:12312
<class 'int'>
Enter a value:1.2323
<class 'float'>


6. Take two numbers and print their sum, difference, and product.

In [ ]:
a,b= input("Enter two numbers (sperate with spaces): ").split()
a =int(a)
b =int(b)
print("sum  ", a+b)
print("difference  ", a-b)
print("product  ", a*b)

Enter two numbers (sperate with spaces): 12312 12
sum   12324
difference   12300
product   147744


7. Convert a string number to an integer and perform addition.

In [ ]:
a = "10"
a = int(a)
a += 10
print (a)

20


8. Take temperature in Celsius and convert it to Fahrenheit.

In [ ]:
temp = int(input("Enter the temperature in celsius: "))
tempf = (temp * (9/5))+ 32
print ("temperature in fahrenheit: ", tempf)

Enter the temperature in celsius: 45
temperature in fahrenheit:  113.0


9. Find the area of a circle from the radius input by user.

In [ ]:
radius = float(input("Enter the radius of the circle: "))
area = 3.14 * (radius ** 2)
print ("area of the circle: ", area)

Enter the radius of the circle: 24
area of the circle:  1808.64


10. Print a sentence using f-strings that includes variables.

In [ ]:
age = 10
name = "ali"
print (f"my name is {name} and my age is {age}")

my name is ali and my age is 10


Operators & Expressions




11. Create a simple calculator for +, -, *, and /.

In [ ]:
a = int(input("Enter 1st Number: "))
b = int(input("Enter 2nd Number: "))
op = input("Enter the operator + , - , * , / ")

if op == '+':
  print(f"{a} + {b} = {a+b}")
elif op == '-':
  print(f"{a} - {b} = {a-b}")
elif op == '*':
  print(f"{a} * {b} = {a*b}")
elif op == '/' and b != 0:
  print(f"{a} / {b} = {a/b}")
elif op == '/' and b == 0:
  print("Division by zero is not possible")
else:
  print("Invalid Operator")


Enter 1st Number: 3
Enter 2nd Number: 2
Enter the operator + , - , * , / *
3 - 2 = 6


12. Check if a number is even or odd using modulus.

In [ ]:
a = int(input("Enter A number"))
if a % 2 == 0:
  print("even ")
else:
  print("odd")

Enter A number4
even 


13. Take birth year as input and calculate age.

In [ ]:
yob = int(input("Enter your year of birth: "))
age = 2025 - yob
print(f"your age is {age}")

Enter your year of birth: 2003
your age is 22


14. Write a program to compute simple interest.

In [ ]:
P = float(input("Enter the principal amount: "))
R = float(input("Enter the rate of interest: "))
T = float(input("Enter the time period in years: "))

SI = (P * R * T) / 100
print(f"Simple Interest with given amount {P}, Rate of interest {R}, and time {T} is {SI}")

Enter the principal amount: 1500
Enter the rate of interest: 5
Enter the time period in years: 10
Simple Interest with given amount 1500.0, Rate of interest 5.0, and time 10.0 is 750.0


15. Calculate BMI from user weight and height.

In [ ]:
w = float(input("Enter your weight in kg: "))
h = float(input("Enter your height in meters: "))

bmi = w / (h ** 2)
bmi *= 100
bmi = round(bmi, 2)
print(f"Your BMI is {bmi}")

Enter your weight in kg: 91
Enter your height in meters: 17
Your BMI is 31.49


16. Use compound assignment operators in a demo program.

In [ ]:
a = 5
a += 5
print (a)
a -=2
print (a)
a *= 2
print (a)
a /= 4
print (a)

10
8
16
4.0


17. Take a number and square it using `**` operator.


In [ ]:
num = float(input("Enter a number: "))
square = num ** 2
print(f"The square of {num} is {square}")

Enter a number: 5
The square of 5.0 is 25.0


18. Demonstrate usage of all comparison operators.


In [ ]:
a , b = 5 ,6
if a < b:
  print (f"{a} is less than {b}")
if a <= b:
  print (f"{a} is less than or equal to {b}")
if a > b:
  print (f"{a} is greater than {b}")
if a >= b:
  print (f"{a} is greater than or equal to {b}")
if a==b:
  print (f"{a} is equal to {b}")
if a!=b:
  print (f"{a} is not equal to {b}")


5 is less than 6
5 is less than or equal to 6
5 is not equal to 6


19. Write a program using logical operators (`and`, `or`, `not`).


In [ ]:
temp = True
smoke = False
smell = True

if temp == True and smoke==True and smell==True:
  print("Alert! Fire")
elif (temp == True or smell==True ) and smoke==False:
  print("No Fire Yet")
elif smoke == True and smell == True:
  print("Alert! Fire")
elif (smoke == False and smell == False) or temp == True:
  print("No! Fire")

print(not(True))
print(not(False))
print(True and False)
print(True and True)
print(True or False)
print(False or False)

No Fire Yet
False
True
False
True
True
False


20. Calculate the area and perimeter of a rectangle.

In [ ]:
length = float(input("Enter the length of the rectangle: "))
width = float(input("Enter the width of the Rectangle" ))

a_o_r = length * width
parameter = (2 * length) + (2 * width)
print(f"Area of Rectangle: {a_o_r}")
print(f"Parameter of Rectangle: {parameter}")

Enter the length of the rectangle: 10 
Enter the width of the Rectangle12
Area of Rectangle: 120.0
Parameter of Rectangle: 44.0



Conditionals
🔢 Q21. Grade Calculator with Validation
Topic: Nested if-else, input validation
Task:
Take marks in 5 subjects (0–100 each), calculate percentage, and assign grade:

A+: ≥90

A: 80–89

B: 70–79

C: 60–69

D: 50–59

F: <50

Also check if any marks are invalid (e.g., >100 or <0) and print "Invalid input".


In [ ]:
Marks =input("Enter Your Marks in English,Maths,Science,Arts,Technology (add space in between)")
Marks = Marks.split()
n_o_s = len(Marks)
if n_o_s != 5:
  print("Invalid Input, Please Enter Marks For 5 Subjects Only.")
  exit()
range_flag = True
for i in range(len(Marks)):
  Marks[i]=int(Marks[i])
  if (Marks[i] < 0) or (Marks[i] > 100):
    range_flag = False
    break
if range_flag == False:
  print("Invalid Input, Please Enter Marks Between 0 and 100 Only.")
  exit()

Avg = sum(Marks)/n_o_s
print(Avg)

if Avg >= 90:
  print("Grade: A+")
elif Avg <= 89 and Avg >= 80:
  print("Grade: A")
elif Avg <= 79 and Avg >=70:
  print("Grade: B")
elif Avg <= 69 and Avg >=60:
  print("Grade: C")
elif Avg <= 59 and Avg >=50:
  print("Grade: D")
elif Avg < 50:
  print("Grade: F")





Enter Your Marks in English,Maths,Science,Arts,Technology (add space in between)100 56 87 93 77
82.6
Grade: A



🌍 Q22. Leap Year Checker
Topic: Nested conditionals
Task:
Take a year as input and check if it's a leap year using proper rules:

Divisible by 4 → leap

But if divisible by 100, then must also be divisible by 400.

Output: “Leap year” or “Not a leap year”




In [ ]:
year = int(input("Enter a year!:  "))
if year % 4 ==0:
  if year % 100 == 0:
    if year % 400 == 0:
      print("Leap Year!")
    else:
      print("Not a lEap Year!")
  else:
    print("leap year")
else:
  print("Not a leap year")

Enter a year!:  2008
leap year


🥇 Q23. Find Maximum of Three Numbers
Topic: if-elif-else
Task:
Take 3 numbers as input and print the largest one.

Bonus: If all are equal, print “All numbers are equal”.





In [ ]:
numbers = [0] * 3
numbers = input("Enter 3 numbers").split()
for i in range(len(numbers)):
  numbers[i] = int(numbers[i])
if numbers[0] == numbers[1] == numbers[2]:
  print("All numbers are equal")
elif numbers[0] >= numbers[1] and numbers[0] >= numbers[2]:
  print(numbers[0])
elif numbers[1]>= numbers[0] and numbers[1] >= numbers[2]:
  print(numbers[1])
elif numbers[2]>= numbers[0] and numbers[2] >= numbers[1]:
  print(numbers[2])

Enter 3 numbers5 5 5
All numbers are equal



➕➖ Q24. Positive, Negative or Zero
Topic: Basic if-elif-else
Task:
Take a number from the user and determine if it’s:

Positive

Negative

Zero

Bonus: Display “Neutral” for zero.


In [ ]:
num = int(input("Enter a number: "))
if num < 0:
  print("Negative")
elif num > 0:
  print("Positive")
else:
  print("Neutral")

Enter a number: 5
Positive



✊✋✌️ Q25. Rock-Paper-Scissors Game (2 Players)
Topic: Multi-condition logic
Task:
Take input from Player 1 and Player 2.
Each enters: rock, paper, or scissors.
Use conditionals to print who wins:

Rock beats scissors

Paper beats rock

Scissors beats paper

If both choose the same, it's a tie.


In [ ]:
p1 = input("Player 1 Turn: Enter R -> Rock, P -> Paper, S -> Scissor ").upper()
if p1 != 'R' and p1 != 'P' and p1 != 'S':
  print("Invalid Input")
  exit()
from IPython.display import clear_output
clear_output()
p2 = input("Player 2 Turn: Enter R -> Rock, P -> Paper, S -> Scissor ").upper()
clear_output()
if p2 != 'R' and p2 != 'P' and p2 != 'S':
  print("Invalid Input")
  exit()
if p1 == p2:
  print("it is a tie!")
elif p1 == 'R' and p2 == 'S':
  print("p1 wins")
elif p1 == 'S' and p2 == 'R':
  print("p2 wins")
elif p1 == 'P' and p2 == 'R':
  print("p1 wins")
elif p1 == 'R' and p2 == 'P':
  print("p2 wins")
elif p1 == 'S' and p2 == 'P':
  print("p1 wins")
elif p1 == 'P' and p2 == 'S':
  print("p2 wins")
else:
  print("Invalid Input")

it is a tie!



🔤 Q26. Vowel or Consonant Checker
Topic: Char input and logic
Task:
Ask the user to input a single alphabet character.
Check:

Is it a letter?

Is it a vowel (a, e, i, o, u)?

If not vowel, print "Consonant"

If not a letter, print "Invalid character"


In [ ]:
vowels = "aeiou"
ch = input("Enter a alphabet character: ").lower()
if ch in vowels:
  print("Vowel")
elif ch.isalpha():
  print("Consonant")
else:
  print("Invalid Character")

Enter a character: 2
Invalid Character



🗳️ Q27. Voting Eligibility Checker
Topic: Simple conditionals
Task:
Ask for user’s age and citizenship status (yes/no).
To vote:

Age must be ≥18

Must be a citizen

Output: “Eligible to vote” or “Not eligible”


In [ ]:
age = int(input("Enter your Age: "))
citizen = input("Citzen or Not (Yes/No)").lower()
if age>=18 and citizen == "yes":
  print("Eligible to vote")
else:
  print("Not Eligible")

Enter your Age: 22
Citzen or Not (Yes/No)yes
Eligible to vote



🔢 Q28. Divisible by Both 5 and 11
Topic: Compound condition
Task:
Take a number and check:

If divisible by 5 and 11 → print “Yes”

Else → print “No”

Bonus: Also show which ones it is divisible by (if only one)


In [ ]:
num = int(input("Eneter a Number: "))
if num % 5 ==0 and num % 11 == 0:
  print("Yes")
elif num%5 == 0 and num%11 !=0:
  print("Yes, Divisible by 5")
elif num%5 != 0 and num%11 ==0:
  print("Yes, Divisible by 11")
else:
  print("No")


Eneter a Number: 32
No



🧒🧑🧓 Q29. Age Category Assigner
Topic: Range-based logic
Task:
Input age → classify into:

Child: <13

Teen: 13–19

Young Adult: 20–35

Adult: 36–59

Senior: ≥60


In [ ]:
age = int(input("Enter Age:  "))
if age in range(13):
  print("child")
elif age in range(13,20):
  print("Teen")
elif age in range(20,36):
  print("Young Adult")
elif age in range(36,60):
  print("Adult")
elif age >= 60:
  print("Senior")

Enter Age:  59
Adult



📋 Q30. Menu-Driven Program with If-Elif-Else
Topic: User choice-based branching
Task:
Display this menu:

1. Add two numbers
2. Find square of a number
3. Convert Celsius to Fahrenheit
4. Exit
Take input from user and perform the task.

Print "Invalid choice" for wrong input.
Add another input to return to menu: simulate it one time only (no loops).

In [ ]:
print("\t1. Add two numbers \n\t2. Find square of a number\n\t3. Convert Celsius to Fahrenheit \n\t4. Exit Take input from user and perform the task \n")
inp = int(input("Select a number:"))
if inp == 1:
  a = int(input("Enter a number"))
  b = int(input("Enter a number"))
  print(f"{a} + {b} = {a+b}")
elif inp ==2:
    a = int(input("Enter a number"))
    print(f" {a} squared = {a**2}")
elif inp == 3:
  temp = float(input("Enter the temperature in celsius: "))
  tempf = (temp * (9/5))+ 32
  print (f"{temp}'C  in fahrenheit: {tempf}'F ")
elif inp == 4:
  print("Exit")
else:
  print("Invalid Choice")


	1. Add two numbers 
	2. Find square of a number
	3. Convert Celsius to Fahrenheit 
	4. Exit Take input from user and perform the task 

Select a number:3
Enter the temperature in celsius: 45
45.0'C  in fahrenheit: 113.0'F 



Loops
Q31. Print Numbers from 1 to N (Inclusive) with Input Validation
Topic: for loop, input validation
Task:
Write a program that takes an integer N as input and prints numbers from 1 to N.

Ensure that the input is a positive integer.

If the input is invalid (e.g., 0 or negative), display "Invalid input. Please enter a positive integer."

Use a for loop to print each number on a separate line.


In [ ]:
N = int(input("Enter a Number"))
if N <= 0:
  print("Invalid Input Please Enter a Positive Integer")
else:
  for i in range(1,N+1):
    print(i)


Enter a Number91
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91



Q32. Countdown from N to 1 with Delay Option
Topic: while loop, input validation
Task:
Take a positive integer N and print numbers from N down to 1.

Ask the user if they want a delay between each print (Y/N).

If yes, wait 0.5 seconds between each number (use time.sleep(0.5)).

Import the time module accordingly.

Validate the input to ensure N is positive.

















In [ ]:
import time
N=int(input("Enter a Number"))

if N <= 0:
  print("Invalid Input Please Enter a Positive Integer")
else:
  delay = input("Do you want a delay between each print (Y/N)").upper()
  for i in range(N,0,-1):
    print(i)
    if delay == "Y":
      time.sleep(0.5)

Enter a Number3
Do you want a delay between each print (Y/N)y
3
2
1


Q33. Calculate Sum of First N Natural Numbers (1 + 2 + … + N)
Topic: for loop, accumulator pattern
Task:
Write a program that:

Accepts a positive integer N.

Calculates the sum of the first N natural numbers using a loop (not formula).

Displays the full expression as well as the result.
Example:
1 + 2 + 3 + 4 + 5 = 15

In [ ]:
N=int(input("Enter a Number: "))
expression = ""
total = 0
if N <= 0:
  print("Invalid Input Please Enter a Positive Integer")
  exit()
for i in range(1,N+1):
  expression = expression + " + " +str(i)
  total = total + i
expression = expression[2:]
print(expression, end = " ")
print(f"= {total}")


Enter a Number: 5
 1 + 2 + 3 + 4 + 5 = 15


Q34. Find Factorial of a Number with Stepwise Output
Topic: Loop, accumulator pattern
Task:
Take a non-negative integer N and compute its factorial using a loop.

If N = 0, the factorial is 1.

Also display the multiplication steps used.
Example:
5! = 5 × 4 × 3 × 2 × 1 = 120

In [ ]:
N =int(input("Enter a Number:  "))

expression = f"{str(N)}! = "
factorial = 1
if N == 0:
  print(f"0! = {factorial}")
  exit()
elif N < 0:
  print("Invalid Input, Please Enter a positive integer. ")
  exit()
else:
  for i in range(N,1,-1):
    expression = expression +  str(i) + " x "
    factorial *= i
  expression = expression + "1"
  expression = expression + f" = {factorial}"
  print(expression)

Enter a Number5
5! = 5 x 4 x 3 x 2 x 1 = 120


Q35. Display Multiplication Table Up to a Limit
Topic: for loop, nested logic
Task:
Ask the user for:

A number num for which the multiplication table is to be generated.

A limit up to which it should be printed.

Example:
If input is num = 7 and limit = 5, output should be:


7 × 1 = 7  
7 × 2 = 14  
7 × 3 = 21  
7 × 4 = 28  
7 × 5 = 35

In [ ]:
import sys
num = int(input("Enter for Table: "))
if num <= 0:
  print("Invalid Input, Please Enter a positive integer. ")
  sys.exit()
limit = int(input("Limit of Table: "))
if limit <= 0:
  print("Invalid Input, Please Enter a positive integer. ")
  sys.exit()

for i in range (1,limit+1):
  print(f"{num} x {i} = {num*i}")

Enter for Table: 5
Limit of Table: 5
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25


Q36. Count Digits in a Number Without String Conversion
Topic: while loop, integer operations
Task:
Write a program to count how many digits are in a given positive integer.

You must not convert the number to a string.

Display the count along with a message.
Example:
Input: 54789 → Output: 5 digits

In [ ]:
num = int(input("Enter a Number: "))
num1 = num
counter = 0
if num == 0:
  counter = 1
  print(f"Input: {num1} --> Output: {counter} digits")
  sys.exit()
while num >= 1:
  num = num//10
  print(num)
  counter += 1
print(f"Input: {num1} --> Output: {counter} digits")


Enter a Number: 5678
567
56
5
0
Input: 5678 --> Output: 4 digits


Q37. Reverse a Number Using Loop (No String Conversion)
Topic: while loop, number manipulation
Task:
Take a positive integer and reverse it using a loop.

Do not use string slicing or conversion.

Display the reversed number.
Example:
Input: 1234 → Output: 4321

In [ ]:
num = int(input("Enter a Number: "))
n = num
num_r = 0
while num > 0:
  digit = num % 10
  num_r = num_r * 10 + digit
  num = num // 10
print(f"Input: {n} --> Output: {num_r}")

Enter a Number: 12345677654321234567654321234567
Input: 12345677654321234567654321234567 --> Output: 76543212345676543212345677654321



Q38. Palindrome Number Checker Without Strings
Topic: Looping, mathematical manipulation
Task:
Check if a given positive integer is a palindrome.

Do not convert to string.

A palindrome reads the same forward and backward (e.g., 121, 1331).
Output:

"Palindrome" if it is,

"Not a palindrome" if not.

In [ ]:
num = int(input("Enter a Number: "))
n = num
num_r = 0
while num > 0:
  digit = num % 10
  num_r = num_r * 10 + digit
  num = num // 10
if num_r == n:
  print(f"input: {n} --> output: {num_r} is a palindrome")
elif num_r != n:
  print(f"input: {n} --> output: {num_r} is not a palindrome")

Enter a Number: 1976
input: 1976 --> output: 6791 is not a palindrome


Q39. Print Even and Odd Numbers Separately Up to N
Topic: Loop, conditionals
Task:
Input a positive integer N and print:

All even numbers from 1 to N on one line.

All odd numbers from 1 to N on another line.

Output Example:

Even: 2 4 6 8  
Odd: 1 3 5 7 9


In [ ]:
num = int(input("Enter a positive integer: "))
if num <= 0:
  print("Invalid Input!")
  sys.exit()
else:
  print("\nOdd Numbers in range\n")
  for i in range(1,num+1,2):
    print(i ,sep =" ", end= " ")
  print("\n\nEven Numbers in range\n")
  for i in range(2,num+1,2):
    print(i ,sep =" ", end= " ")



Enter a positive integer: 10

Odd Numbers in range

1 3 5 7 9 

Even Numbers in range

2 4 6 8 10 

Q40. Print First N Fibonacci Numbers with Explanation
Topic: Loop, sequence generation
Task:
Take a positive integer N and print the first N terms of the Fibonacci sequence.

The sequence starts with 0 1, and every term is the sum of the previous two.

Print a note explaining the logic to the user before the output.

Example:


Fibonacci Sequence (first 6 terms):  
0 1 1 2 3 5

In [ ]:
num = int(input("Enter a positive integer: "))
if num <= 0:
  print("Invalid Input!")
  sys.exit()
else:
  print("Explanation \nWe Start with a number taken as input from user and use it as a limit for while loop. we create a list fibonacci, with first two elements 0, 1 already in")
  print("then we start the loop which iterates untill num becomes less then i, i starts from 2, we keep adding into the list, the sum of last two indices. lastly we print ")
  print("the series one by one.\n\n")
  i = 2
  a = 0
  b = 1
  fibonacci = [a,b]
  print(fibonacci[0], end = " ")
  print(fibonacci[1], end = " ")
  while i < num :
    fibonacci.append( fibonacci[i-1] + fibonacci[i-2])
    print(fibonacci[i], end = " ")
    i+=1


Enter a positive integer: 10
Explanation 
We Start with a number taken as input from user and use it as a limit for while loop. we create a list fibonacci, with first two elements 0, 1 already in
then we start the loop which iterates untill num becomes less then i, i starts from 2, we keep adding into the list, the sum of last two indices. lastly we print 
the series one by one.


0 1 1 2 3 5 8 13 21 34 

## Combined Logic (Loops + Conditionals)




41. FizzBuzz Challenge (1 to 100):
Write a program that prints numbers from 1 to 100. For multiples of 3, print "Fizz" instead of the number. For multiples of 5, print "Buzz". For numbers divisible by both 3 and 5, print "FizzBuzz". All others should print as-is.


In [ ]:
for i in range(1,101):
  if i % 5 == 0 and i % 3 == 0:
    print("FizzBuzz")
  elif i % 5 == 0:
    print("Buzz")
  elif i % 3 == 0:
    print("Fizz")
  else:
    print(i)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz
Buzz



42. Prime Number Checker:
Accept a number from the user and check whether it is a prime number or not. Make sure your solution handles edge cases like 0, 1, and negative numbers with appropriate messaging.


In [ ]:
import sys

num =int(input("Enter a Positive Integer: "))
check = False
if num <= 0:
  print("Invalid Input!")
elif num==1:
  print("not prime.")
else:
  for i in range(2,10):
    if num % i == 0 and num != i:
      check = False
      break
    else:
      check = True
  print("Prime") if check == True else print("Not Prime")



Enter a Positive Integer: 5
Prime



43. All Primes in a Range:
Take two integers as input and display all prime numbers between them (inclusive). Display the total number of prime numbers found in that range.



In [ ]:
rep = "yes"
while rep == "yes":
  start = int(input("Enter the starting number: "))
  end = int(input("Enter the ending number: "))
  if start <= 0 or end <= 0:
    print("Invalid Input! Positive integers only.")
    rep = input("Wanna Enter Again! yes/no: ").lower()
  else:
    rep = "no"

counter = 0

# Ascending Range
if start < end:
  for i in range(start, end + 1):
    if i == 1:
      continue
    check = True
    for j in range(2, int(i**0.5) + 1):
      if i % j == 0:
        check = False
        break
    if check:
      print(i, end=" ")
      counter += 1

# Descending Range
else:
  for i in range(start, end - 1, -1):
    if i == 1:
      continue
    check = True
    for j in range(2, int(i**0.5) + 1):
      if i % j == 0:
        check = False
        break
    if check:
      print(i, end=" ")
      counter += 1

print(f"\nTotal Prime Numbers: {counter}")


Enter the starting number: 100
Enter the ending number: 10
97 89 83 79 73 71 67 61 59 53 47 43 41 37 31 29 23 19 17 13 11 
Total Prime Numbers: 21


44. Armstrong Number Checker:
Check if a given number is an Armstrong number. A number is an Armstrong number if the sum of its digits raised to the power of the number of digits is equal to the original number. Example: 153 → 1^3 + 5^3 + 3^3 = 153.


In [ ]:
num = int(input("Enter a number:  "))
temp = num
sum = 0
digit = []
while num != 0:
  digit.append(num % 10)
  num = num // 10
power = len(digit)
for i in range(power):
  sum += (digit[i] ** power)
if temp == sum:
  print(f"Armstrong Number:  given: {temp} == calculated: {sum}")
else:
  print(f"Not Armstrong Number:  given: {temp} != calculated: {sum}")

Enter a number:  153
Armstrong Number:  given: 153 == calculated: 153



45. Sum of Even Digits in a Number:
Take a positive integer as input and calculate the sum of only the even digits present in the number. Display both the digits and the final sum.


In [ ]:
num = int(input("Enter A Number: "))
digit = []
temp = num
sum = 0
even = []
while num != 0:
  digit.append(num % 10)
  num = num // 10
length = len(digit)
for i in range(length):
  if digit[i] % 2 == 0:
    sum += digit[i]
    even.append(digit[i])

  else:
    continue
print(f"Sum of even numbers in {temp} == {sum}")
print(even)

Enter A Number: 23456
Sum of even numbers in 23456 == 12
[6, 4, 2]



46. Star Pattern: Right-Angled Triangle (using *):
Ask the user for the number of rows and display a right-angled triangle pattern using stars. The triangle should be left-aligned.


In [ ]:
num = int(input("Enter number of rows"))

for i in range(1,num+1):
  print()
  for j in range(i):
    print("*", end = " ")

Enter number of rows5

* 
* * 
* * * 
* * * * 
* * * * * 


47. Star Pattern: Inverted Right-Angled Triangle:
Take input for rows and display an inverted triangle made of stars aligned to the left, starting from the largest row count.



In [ ]:
num = int(input("Enter number of rows"))



for i in range(num,0,-1):
  print()
  for j in range(i):
    print("*", end = " ")


Enter number of rows5

* * * * * 
* * * * 
* * * 
* * 
* 

48. Number Pyramid Pattern:
Create a number pyramid where each row i contains i repeated numbers. For example, row 3 should display 333. The number of rows should be taken as input.


In [ ]:
num = int(input("Enter number of rows"))

for i in range(num+1):
  print()
  for j in range(num-i):
    print(" ",end = " ")
  for j in range(i): # complete visually accurate pyramid if for j in range((2*i)-1)
    print(i, end= " ")

Enter number of rows5

          
        1 
      2 2 
    3 3 3 
  4 4 4 4 
5 5 5 5 5 


49. Factor Finder with Classification:
Take a number as input and print all of its factors. Also display how many of them are even and how many are odd.



In [ ]:
num = int(input("Enter number:  "))
if num <0:
    print("Invalid Input! positive Integers Only")
    sys.exit()


factors = []
for i in range(1, num+1):
  if num % i == 0:
    factors.append(i)
length = len(factors)
even = []
odd = []
for i in range(length):
  print(factors[i], end=" ")
even_count = 0
odd_count = 0
for factor in factors:
    if factor % 2 == 0:
        even_count += 1
    else:
        odd_count += 1

print(f"\nEven Factors: {even_count} | Odd Factors: {odd_count}")

Enter number:  10
1 2 5 10 
Even Factors: 2 | Odd Factors: 2


50. GCD (HCF) Calculator for Two Numbers:
Ask the user to input two positive integers and find their greatest common divisor (GCD), also known as highest common factor (HCF). Do not use any built-in GCD functions.

In [ ]:
a = int(input("Enter a : "))
b = int(input("Enter b : "))

result = min(a, b)
while result > 0:
  if a % result == 0 and b % result == 0:
    break
  result -= 1
print(result)

Enter a : 24
Enter b : 36
12


## Strings

51. Count the number of vowels and consonants in a given string.

Ignore digits and special characters.

Treat both uppercase and lowercase letters as valid.


In [ ]:
vowel = "aeiou"
given = input("Enter a string: ")
v_count=0
c_count=0
vowels = []
constants = []
for ch in given:
  if ch.lower() in vowel:
    v_count += 1
    vowels.append(ch)
  elif ch.isalpha():
    c_count += 1
    constants.append(ch)
  else:
    continue

print(f"number of vowels:  {v_count}    {vowels}")
print(f"number of constants:  {c_count}    {constants}")


Enter a string: this is a string 123123123./..'.;.
number of vowels:  4    ['i', 'i', 'a', 'i']
number of constants:  9    ['t', 'h', 's', 's', 's', 't', 'r', 'n', 'g']


52. Check if a string is a palindrome.

Case-insensitive.

Ignore spaces while checking.

In [ ]:
given = input("Enter a string: ").lower()

straight =[]
reverse = []
for ch in given:
  if ch != " ":
    straight.append(ch)
reverse = straight[::-1]
if reverse == straight:
  print("Palindrome")
else:
  print("Not Palindrome")

Enter a string: 111 tht111
Palindrome


53. Count how many words are in a sentence.

A word is any sequence of non-space characters separated by one or more spaces.

Don't use .split().


In [ ]:
given = input("Enter a string: ").lower().strip()
words = []
word = ""
length = len(given)
i = 0
while i <= length-1:
  if given[i] != " ":
    word = word + given[i]
    i += 1
  else:

    words.append(word)
    word = ""
    i+=1
if word:
  words.append(word)
now= len(words)
print(f"Number of words : {now}     {words}")

Enter a string: this is a string
Number of words : 4     ['this', 'is', 'a', 'string']


54. Replace all spaces in a string with hyphens (-).

Avoid using the built-in .replace() function.

In [ ]:
given = input("Enter a string: ").lower().strip()
new = ""

for ch in given:
  if ch != " ":
    new = new + ch
  else:
    new = new +"-"

print(f"Output: {new}")

Enter a string: this is a string 
Output:  this-is-a-string


55. Convert the first letter of each word to uppercase.

Don’t use .title() or .capitalize().

In [ ]:
given = input("Enter a string: ").lower().strip()
words = ""
word = ""
i = 0
length = len(given)
while i <= length-1:
  if given[i] != " ":
    word = word + given[i]
    i += 1
  else:
    word = word[0].upper() + word[1:]
    words = words + " " + word
    word = ""
    i+=1
if word:
  word = word[0].upper() + word[1:]
  words = words + " " + word

print(f"New : {words.strip()}")

Enter a string: this is a string
New :  This Is A String


56. Remove all digits from a string.

Output the cleaned string with only alphabetic characters and symbols.

In [ ]:
given = input("Enter a String: ")
clean = ""
for ch in given:
  if ch.isdigit():
    clean = clean + ""
  else:
    clean = clean + ch
print(f"cleaned String : \n {clean}")

Enter a Stringthis is a string 1with 2nu23mb4er5s
cleaned String : 
 this is a string with numbers



57. Count how many times each character appears in a string.

Exclude spaces from the count.

Output must show: character → count (in any order).

In [ ]:
given = input("Enter a String: ").lower().strip()
length = len(given)
count = 0
done = ""
for ch in given:
  count = 0
  if ch not in done:
    for i in range(length):
      if ch == given[i] and ch != " ":
        count += 1
      elif ch == " ":
        continue
  else:
    continue
  done = done + ch
  if ch != " ":
    print(f"{ch} --> {count}")






Enter a String: This is a string
t --> 2
h --> 1
i --> 3
s --> 3
a --> 1
r --> 1
n --> 1
g --> 1


58. Remove all duplicate characters from a string while keeping order.

For example, "aabbccdd" → "abcd".

In [ ]:
given = input("Enter a String: ").lower().strip()
final = ""
for i in range(len(given)):
  if given[i] in final:
    continue
  else:
      final = final + given[i]

print(final)


Enter a String: aaaabbccccdddd
abcd


59. Find the longest word in a sentence.

If multiple have the same length, return the first one.

No built-in .split().

In [ ]:
given = input("Enter a String: ").lower().strip()
word = []
longest = []
i = 0
while i < len(given):

  if given[i] != " ":# run untill encounter a space
    word.append(given[i])
    i+=1

  else: #only runs when a word has been completly added to the list

    if len(word)>len(longest):# checks if current longest is shorter than the current word
      longest = word
      i+=1
      word=[]# remove the previous word

    else: # if current longest is short, just remove it and move on with the loip.
      i+=1
      word=[] # remove the previous word
      continue
# loop ends here

if len(word) > len(longest): # compare last word with the longest word.
    longest = word

l_word = "" # convert back into a word from a list
for i in range(len(longest)):
  l_word = l_word + longest[i]
print(l_word)






Enter a String: this is the longest wordeverwriiten
wordeverwriiten



60. Reverse the order of words in a sentence.

"Hello World Python" → "Python World Hello"

No .split() or reversed() — use manual logic.

In [ ]:
given = input("Enter a String: ").lower().strip()
word = []
sentence = []
i = 0
while i < len(given):

  if given[i] != " ":# run untill encounter a space
    word.append(given[i])
    i+=1

  else: # when space encounter add whole word into a list -> sentence
    sentence.append("".join(word)) # this makes sure, each word added is joined, means not different characters but a word, followed by a space.
    word = [] # reset the word list to empty for new word.
    i+=1

if word: # add the word after last space.
  sentence.append("".join(word))
sentence = sentence[::-1] # reverse list of words not characters

l_word = " ".join(sentence) # join into a string with added space between.
print(l_word)






Enter a String: this is a string
string a is this


## Lists


61. Find the Second Largest Unique Element in a List

 Input: [4, 1, 6, 7, 4, 7, 6]

 Output: 6

 Hint: No built-in sort. Do it manually.

In [ ]:
import sys

lis = input("Enter the numbers:  ").split()
lis =[ int(i) for i in lis ]

if len(lis) <= 1:
  print("Not Enough Values")
  sys.exit()

#remove repeated
lis = list(set(lis))

if len(lis) <= 1 :
  print("No Second Largest, only one value.")
  sys.exit()


temp = 0

for i in range(len(lis)):
  for j in range(i+1,len(lis)):
    if lis[i] < lis[j]:
      temp = lis[i]
      lis[i] = lis[j]
      lis[j] = temp

print(f"from given {lis} \nthe second largest is {lis[1]}")


Enter the numbers:  55 66 77 33 22 88 99
from given [99, 88, 77, 66, 55, 33, 22] 
the second largest is 88


62. Count Frequency of Each Character in a String

 Input: "banana"

 Output: {'b': 1, 'a': 3, 'n': 2}

In [ ]:
given = input("Enter a string: ").strip()
orignal = []
output = "{"
count = 0
for ch in given:
  if ch != " ":
    orignal.append(ch)

unique = list(set(orignal))
print(unique)


for i in range(len(unique)):
  count = 0
  for j in range(len(orignal)):
    if unique[i] == orignal[j]:
      count+=1
  output += f"'{unique[i]}': {count},"


output = output[:-1]
output += "}"
print(output)



Enter a string: banana
['b', 'a', 'n']
{'b': 1,'a': 3,'n': 2}


63. Find All Prime Numbers in a List

 Input: [3, 4, 5, 6, 7, 8, 9, 10]

 Output: [3, 5, 7]

In [ ]:
all = input("Enter a list: ").split()
all = list(set([int(i) for i in all]))
prime = []

for i in all:
  if i == 1 or i == 0:
    continue
  check = True
  for j in range(2, int(i**0.5) + 1):
    if i % j == 0:
      check = False
      break
  if check == True:
    if i not in prime:
      prime.append(i)

print(prime)

Enter a list: 0 1 2 3 4 5 6
[2, 3, 5]


64. Reverse a List Without Using Reverse or Slicing

 Input: [1, 2, 3, 4]

 Output: [4, 3, 2, 1]

In [ ]:
all = input("Enter a list: ").split()
all = [int(i) for i in all]
rev = []
for i in range(len(all)-1,-1,-1):
  rev.append(all[i])
print(rev)

Enter a list: 11 12 13 14 24 33
[33, 24, 14, 13, 12, 11]


65. Find the Most Frequent Word in a Sentence

 Input: "cat bat cat bat cat rat bat"

 Output: cat (or bat, depending on tie-breaking logic)

In [ ]:
l = input("enter a string: ").split()
unique = list(set(l))
count= ['',0]
cnt = 0
for i in unique:
  for j in l:
    if i==j:
      cnt += 1
      if cnt > count[1] or (cnt == count[1] and l.index(i) < l.index(count[0])):
        count[0] = i
        count[1] = cnt
  cnt = 0

print(count[0], count[1])


enter a string: cat bat cat bat cat rat bat
cat 3


66. Remove All Duplicates from a List Without Using Set

 Input: [1, 2, 2, 3, 4, 4, 4, 5]

 Output: [1, 2, 3, 4, 5]

In [ ]:
l = input("Enter a list: ").split()
l = [int(i) for i in l]
done = []
for i in l:
  if i not in done:
    done.append(i)
print(done)


Enter a list: 1 2 2 3 4 4 4 5
[1, 2, 3, 4, 5]


67. Check if Two Strings are Anagrams Without Collections

 Input: "listen", "silent"

 Output: True

In [ ]:
tstrings = input("Enter two strings:  ").split()

l1 = [ch for ch in tstrings[0]]
l2 = [ch for ch in tstrings[1]]
if len(l1) == len(l2):
  l1.sort()
  l2.sort()

  if l1 == l2:
    print(True)
  else:
    print(False)
else:
  print(False)

Enter two strings:  malik maaalik
False


68. Rotate a List to the Right by K Steps

 Input: [1, 2, 3, 4, 5], k = 2

 Output: [4, 5, 1, 2, 3]

In [ ]:
l = input("Enter a list: ").split()
l = [int(i) for i in l]
k = int(input("Enter step: "))

l = l[-k:] + l[:-k]
# l = l[k:] + l[:k] for left rotation
print(l)

Enter a list: 1 2 3 4 5
Enter step4
[2, 3, 4, 5, 1]


69. Merge Two Sorted Lists into One Sorted List (Without Using Sort)


 Input: [1, 3, 5], [2, 4, 6]

 Output: [1, 2, 3, 4, 5, 6]



In [ ]:

l = input("Enter a list: ").split()
l = [int(i) for i in l]

j  = input("Enter a list: ").split()
j = [int(i) for i in j]

# would also be correct since i used sorted not sort but
# k = sorted(l[:] + j[:])
# print(k)

k = l[:] + j[:]

for i in range(len(k)):
  for j in range(i+1,len(k)):
    if k[i] > k[j]:
      temp = k[i]
      k[i] = k[j]
      k[j] = temp

print(k)


Enter a list: 1 3 5
Enter a list: 2 4 6
[1, 2, 3, 4, 5, 6]


70. Find the Longest Palindromic Word in a Sentence

 Input: "madam racecar apple level rotor"

 Output: racecar

In [ ]:
words = input("Enter a Sentence: ").split()
longest = ['',0]
for word in words:
  if word == word[::-1]:
    if len(word) > longest[1]:
      longest[0] = word
      longest[1] = len(word)
print(longest)

Enter a Sentence: this is a carr arara ararar
['arara', 5]


## Tuples
71. Tuple Unpacking with Skipped Elements
Given t = (10, 20, 30, 40, 50, 60), unpack it into:

a = 10

b = 20

rest = [30, 40]

c = 50

d = 60
Hint: use unpacking with *.

In [ ]:
t = (10, 20, 30, 40, 50, 60)
rest = []
a , b,*rest, c, d = t
print(a,b,rest,c,d)

10 20 [30, 40] 50 60


72. Immutable But Mutable?

What will be the output of the following and why?

        t = ([1, 2], [3, 4])

        t[0] += [5]

        print(t)

        Answer:

        the answer would be TypeError because tuple are not mutable in structure yes. but whats within a tuple is
        
        mutable or not depends on itself. lets say we have a list, which is itself mutable so this renders the list part of the
        
        tuple mutable, since list is stored in a same address, changing the list doesnt change it's reference.

        if we append to list, that is possible but operatons such as adding to what is existing in the list causes errors

In [ ]:
t = ([1, 2], [3, 4])
t[0].append(4) # this works cuz we are not assiging something new, we are adding to the list
print(t)

t[0] += [5] # this here is chaning what is in the list. so we cant

print(t)

([1, 2, 4], [3, 4])


TypeError: 'tuple' object does not support item assignment

73. Tuple Transformation

        Write a function that takes a tuple of strings and returns a new tuple where

        each string is reversed and only strings longer than 3 characters are included.

        Example:

        Input: ("one", "three", "four", "go")

        Output: ("eerht", "ruof")


In [ ]:
I = tuple(input("Enter a tuple of strings:  ").split())
l= []
for i in I:
  if len(i) > 3:
    l.append(str(i))
  else:
    continue
for i in range(len(l)):
  l[i] = l[i][::-1]

R = tuple(l)
print(R)


Enter a tuple of strings:  THSDH SDHSDHS SDSDH
('HDSHT', 'SHDSHDS', 'HDSDS')


74. Tuple Set Symmetric Difference

        Given two tuples:

        t1 = (1, 2, 3, 4, 5)

        t2 = (4, 5, 6, 7)

        Write a function that returns a tuple of elements that are in either t1 or t2
        but not both.

        Expected Output: (1, 2, 3, 6, 7) — order may vary.

In [ ]:
t1 =  input("Enter first tuple: ").split()
t2 =  input("Enter second tuple: ").split()
t1 = [int(i) for i in t1]
t2 = [int(i) for i in t2]
t1 = tuple(t1)
t2 = tuple(t2)
l = []
for i in t1:
  if i in t1 and i in t2:
    continue
  else:
    l.append(i)

for i in t2:
  if i in t1 and i in t2:
    continue
  else:
    l.append(i)
l = tuple(l)
print(l)

Enter first tuple: 112 123 125 122
Enter second tuple: 122 125 233 235 654
(112, 123, 233, 235, 654)


75. Index-Based Tuple Swapping

        Given a tuple of 10 elements, write a function to return a new tuple where

        elements at even indices are reversed, but odd index elements remain the same.

        Example:

        Input: (10, 11, 20, 21, 30, 31)

        Output: (30, 11, 20, 21, 10, 31)

In [ ]:

t1 = input("Enter first tuple: ").split()
t1 = [int(i) for i in t1]
t1 = tuple(t1)
t2 = []
t2 = list(t1)
l1 = []
for i in range(len(t1)):
  if i%2==0:
    l1.append(t1[i])
  else:
    l1.append(0)

l1 = l1[::-1]
if len(l1)%2 == 0:
  for i in range(len(t2)):
    if i%2==0:
      t2[i] = l1[i+1]
else:
  for i in range(len(t2)):
    if i%2==0:
      t2[i] = l1[i]

print(t2)


# for i in range(1,len(t1)+1,2):




Enter first tuple: 10 11 12 13 14
[14, 11, 12, 13, 10]


## Dictionaries
76. Nested Access Challenge
        Given d = {'a': {'b': {'c': {'d': 100}}}}, write a one-liner to access the value 100.



In [ ]:
d = {'a': {'b': {'c': {'d': 100}}}}
print(d["a"]["b"]["c"]["d"])

100



77. Dictionary from List of Tuples
        Given lst = [("a", 1), ("b", 2), ("a", 3)], convert it into:
        {'a': [1, 3], 'b': [2]}



In [ ]:
lst = [("a", 1), ("b", 2), ("a", 3)]
rec = {}
for i in lst:
  if i[0] not in rec.keys():
    rec[i[0]] = list()
    rec[i[0]].append(i[1])
  else:
    rec[i[0]].append(i[1])
print(rec)


{'a': [1, 3], 'b': [2]}


78. Dictionary with Maximum Value Key
        You are given scores = {'Ali': 90, 'Zara': 96, 'John': 92}
        Print the name of the person with the highest score, without using max(scores) directly.

In [ ]:
scores = {'Ali': 90, 'Zara': 96, 'John': 92,'Kami':96}
max= 0
for i in scores.keys():
  if max < scores[i]:
    max = scores[i]
for k, v in scores.items():
    if max == v:
        print(k)


Zara
Kami




79. Remove Key If Value is List
          Given d = {'a': 1, 'b': [2, 3], 'c': 'text', 'd': [5, 6]}
          Remove all keys whose values are of list type.
          Final Output: {'a': 1, 'c': 'text'}



In [ ]:
d = {'a': 1, 'b': [2, 3], 'c': 'text', 'd': [5, 6]}
l = []
for i in d.keys():
  x = d[i]
  if isinstance(x, list):
    l.append(i)
print(l)
for i in l:
  d.pop(i)
print(d)

['b', 'd']
{'a': 1, 'c': 'text'}




80. Reverse Key-Value Mapping
          Given d = {'a': 1, 'b': 2, 'c': 1}
          Convert it into:
          {1: ['a', 'c'], 2: ['b']}


In [ ]:
d = {'a': 1, 'b': 2, 'c': 1, 'd' : 4,}
n_key = list(set(list(d.values())))
n_vals = []
n_d = {}
j = 0
for i in n_key:
  n_d[i] = list()

for i in d.keys():
  for j in n_key:
    if d[i] == j:
      n_d[j].append(i)

print(n_d)



{1: ['a', 'c'], 2: ['b'], 4: ['d']}



81. Flatten Nested Dictionary
          Input:
          d = { 'name': 'Ali', 'address': {'city': 'Lahore', 'zip': 54000}, 'contact': {'email': 'ali@email.com', 'phone': '12345'} }
          Expected Output:
          { 'name': 'Ali', 'city': 'Lahore', 'zip': 54000, 'email': 'ali@email.com', 'phone': '12345' }


In [ ]:
d = {
    'name': 'Ali',
    'address': {
        'city': 'Lahore',
        'zip': 54000
        },

    'contact': {
        'email': 'ali@email.com',
        'phone': '12345'
        }
    }

n_d = {}
for k,v in d.items():
  if isinstance(v,dict):
    for i in v.keys():
      n_d[i] = v[i]
  else:
    n_d[k] = d[k]



  #
  #   pass
n_d

{'name': 'Ali',
 'city': 'Lahore',
 'zip': 54000,
 'email': 'ali@email.com',
 'phone': '12345'}



            82. Group by First Character
            Given words = ['apple', 'banana', 'cherry', 'apricot', 'blueberry']
            Create a dictionary grouped by the first character:
            { 'a': ['apple', 'apricot'], 'b': ['banana', 'blueberry'], 'c': ['cherry'] }



In [ ]:
first = 0
d = {}
words = ['apple', 'banana', 'cherry', 'apricot', 'blueberry']
for i in words:
  if i[first] not in d.keys():
    d[i[first]] = list()
    d[i[first]].append(i)
  else:
    d[i[first]].append(i)
print(d)



{'a': ['apple', 'apricot'], 'b': ['banana', 'blueberry'], 'c': ['cherry']}


            83. Frequency of Words with Sorting
            Given sentence = "this is a test this is only a test"
            Output word frequency as a list of tuples sorted alphabetically by word:
            [('a', 2), ('is', 2), ('only', 1), ('test', 2), ('this', 2)]


In [ ]:
sentence = "this is a test this is only a test"
l = sentence.split()
d = {}
for i in l:
  if i not in d.keys():
    d[i] = 1
  else:
    d[i]+= 1
l = sorted(list(d.items()))
l

[('a', 2), ('is', 2), ('only', 1), ('test', 2), ('this', 2)]


              84. Merge Two Dicts Without Overwriting
              Given:
              d1 = {'a': 1, 'b': 2}
              d2 = {'b': 3, 'c': 4}
              Expected Output:
              {'a': 1, 'b': [2, 3], 'c': 4}


In [ ]:
d1 = {'a': 1, 'b': 2}
d2 = {'b': 3, 'c': 4}

d = {}
d.update(d1)

for key, value in d2.items():
  if key in d:

    if isinstance(d[key],list):
      d[key].append(value)

    else:
      d[key] = [d[key], value]

  else:
    d[key] = value

print(d)


{'a': 1, 'b': [2, 3], 'c': 4}


              85. Key with Highest Frequency in String
              Given s = "aaabbccccdde"
              Use a dictionary to count the frequency of each character.
              Print the character(s) with the highest frequency and its count.
              Expected Output:
              Character: c
              Frequency: 4

In [ ]:
s = "aaabbccccdde"
d = {}
for ch in s:
  if ch in d.keys():
    d[ch] += 1
  else:
    d[ch] = 1

max = 0
larg = ''
for k,v in d.items():
  if v > max:
    max = v
    larg = k
  else:
    continue
print(f"Character: {larg}")
print(f"Frequency: {max}")


Character: c
Frequency: 4


## Sets

86. Q1. Common Elements Counter
Write a program to find the number of elements common in three sets.

        # Example:
        s1 = {1, 2, 3, 4}
        s2 = {2, 3, 5}
        s3 = {0, 2, 3, 6}

        # Output:
        # Common elements: 2

In [ ]:
s1 = {1, 2, 3, 6}
s2 = {2, 3, 6}
s3 = {0, 2, 3, 6}

common = (s1.intersection(s2)).intersection(s3)
l = list(common)
print(f"Common elements: {len(l)}")

Common elements: 3


87. Symmetric Difference with Multiple Sets

        Given n sets in a list, return the symmetric difference across all sets.
        # Example:
        sets = [{1, 2}, {2, 3}, {3, 4}]
        # Output: {1, 4}

In [ ]:
sets = [{1, 2}, {2, 3}, {3, 4}]
out = set()

for i in sets:
  out.symmetric_difference_update(i)
print(out)

{1, 4}


88. Unique Word Extractor from Sentences
Given a list of strings (sentences), extract a set of all unique words across the sentences.

        # Example:
        sentences = ["this is good", "this is bad", "good and bad"]
        # Output: {'this', 'is', 'good', 'bad', 'and'}

In [ ]:
sentences = ["this is good", "this is bad", "good and bad"]
s = ""
for sentence in sentences:
  for i in sentence:
    print(i)
    s += i
  s += " "
  print(s)
l = s.split()
l = sorted(list(set(l)))
l

t
h
i
s
 
i
s
 
g
o
o
d
this is good 
t
h
i
s
 
i
s
 
b
a
d
this is good this is bad 
g
o
o
d
 
a
n
d
 
b
a
d
this is good this is bad good and bad 


['and', 'bad', 'good', 'is', 'this']

89. Duplicate Remover
Given a list of numbers, print only the unique elements, preserving the original order of first occurrences.
        # Input: [1, 2, 2, 3, 1, 4]
        # Output: [1, 2, 3, 4]

In [ ]:
inn = [1001, 3000, 1000, 10000, 1, 4, 99999]
seen = set()
out = []
for i in inn:
  if i not in seen:
    out.append(i)
    seen.add(i)
print(out)

[1001, 3000, 1000, 10000, 1, 4, 99999]


90. Set Operations Summary
Write a function that takes two sets and returns a dictionary summarizing:

* union
* intersection
* difference (set1 - set2)
* symmetric difference

        # Output:
        # {
        #   "union": {...},
        #   "intersection": {...},
        #   "difference": {...},
        #   "symmetric_diff": {...}
        # }

In [ ]:
def setoperation (a,b):
  dic = {}
  dic["union"] = a.union(b)
  dic["intersection"] = a.intersection(b)
  dic["difference"] = a.difference(b)
  dic["symmetric_diff"] = a.symmetric_difference(b)
  return dic

a= {1,2,3,4}
b= {1,5,6,4}

d = setoperation(a,b)
d

{'union': {1, 2, 3, 4, 5, 6},
 'intersection': {1, 4},
 'difference': {2, 3},
 'symmetric_diff': {2, 3, 5, 6}}

##Functions – 10 Questions (Progressive Difficulty)

91. Basic Calculator
        Create a function that takes two numbers and a string operator (+, -, *, /) and returns the result.


In [ ]:
def cal(num1,num2,op):
  if op not in "+-*/":
    return "Wrong Operator."
  elif op == "+":
    return num1+num2
  elif op == "-":
    return num1-num2
  elif op == "*":
    return num1*num2
  elif op == "/":
    if num2 == 0:
      return "division by zero is not possible"
    else:
      return num1/num2
print(cal(25,5,"/"))
print(cal(25,5,"-"))
print(cal(25,5,"+"))
print(cal(5,5,"*"))
print(cal(5,0,"/"))
print(cal(5,2,"'"))

5.0
20
30
25
division by zero is not possible
Wrong Operator.



92. Count Vowels in String
        Create a function that counts the number of vowels in a given string.


In [ ]:
def count_vowels(string):
  count = 0
  for i in range(len(string)):
    if string[i] in "aeiou":
      count+=1
      print(string[i])
    else:
      continue
  return count
print(count_vowels("vowels"))



o
e
2



93. Factorial (Recursive & Iterative)
          Write two versions of a function to calculate the factorial of a number: one using recursion, one using a loop.


In [ ]:
def l_fac(num):
  fac = 1
  for i in range(1,num+1):
    fac *= i
  return fac

def r_fac(num):
  if num == 0:
    return 1
  else:
    return num * r_fac(num-1)

print("recursion: ",r_fac(6))
print("loop: ",l_fac(6))

recursion:  720
loop:  720




94. Prime Checker
      Write a function that returns True if a number is prime, else False.


In [ ]:
def is_prime(num):

  if num < 2:
    return "False"
  for i in range(2,int(num**0.5)+ 1):
    if num % i == 0:
      return "False"
  return "True"

print(is_prime(78))
print(is_prime(7))
print(is_prime(2))
print(is_prime(1))
print(is_prime(4))
print(is_prime(6))

False
True
True
False
False
False



95. Palindrome Checker
        Write a function that checks whether a given string is a palindrome (ignoring spaces and case).


In [ ]:
def palindrome(s):
  s = s.lower()
  l = []
  for i in s:
    if i == " ":
      continue
    else:
      l.append(i)
  r_l = l[::-1]
  if l == r_l:
    return "Palindrome"
  else:
    return "Not Palindrome"

print(palindrome("racecar"))
print(palindrome("racercar"))



Palindrome
Not Palindrome



96. Find All Divisors
Write a function that returns all the divisors of a number as a list.


        # Input: 12
        # Output: [1, 2, 3, 4, 6, 12]


In [ ]:
def divisor(num):
  div = []
  for i in range(1,num+1):
    if num % i == 0:
      div.append(i)
  return div
divisor(56)


[1, 2, 4, 7, 8, 14, 28, 56]

97. Function with Default Argument
        Write a function greet(name="Guest") that prints a greeting. If no name is provided, it uses "Guest".


In [ ]:
def greet(name="Guest"):
  return(f"Hi ,{name}. Welcome to Python Programming")

print(greet())
print(greet("Ali"))


Hi ,Guest. Welcome to Python Programming
Hi ,Ali. Welcome to Python Programming



98. Variable-Length Arguments (Flexible Sum)
      Create a function flex_sum(*args) that takes any number of numbers and returns their sum.


In [ ]:
def flex_sum(*num):
  sum = 0
  for i in num:
    sum += i
  return sum
print(flex_sum(1))
print(flex_sum(1,2))
print(flex_sum(1,2,3))
print(flex_sum(1,2,3,4))

1
3
6
10



99. Function as Argument
Write a function that takes another function and a list, and applies the function to each element.


      # Example:
      def square(x): return x*x
      apply(square, [1, 2, 3])  # Output: [1, 4, 9]


In [ ]:
def square(x): return x*x
def apply(square,l):
  j = 0
  for i in l:
    l[j] = square(i)
    j+=1
  return l
lis= [1,2,3,4,5]

print(apply(square,lis))


[1, 4, 9, 16, 25]


100. Lambda with Sorting
Given a list of tuples (name, score), sort the list in descending order of score using a lambda function.


        # Example:
        [("Ali", 89), ("Zara", 95), ("John", 75)]
        # Output:
        [("Zara", 95), ("Ali", 89), ("John", 75)]


In [ ]:
l = [("Ali", 89), ("Zara", 95), ("John", 75)]
l = sorted(l, reverse=True, key= lambda x: x[1]) # remember this x here for each tuple in x, and for each x return x[1]
# for each tuple return second elementy of the tuple
l



[('Zara', 95), ('Ali', 89), ('John', 75)]

## OOP


101. BankAccount Class — Basic Structure
Design a class named BankAccount which allows creating a bank account with attributes like account holder name and an initial balance. The class should have instance methods deposit(amount) and withdraw(amount), which respectively add to or subtract from the balance, while preventing withdrawals that would cause a negative balance. Also add a method to display the current balance when called.

    Key Concepts: __init__, self, encapsulation, instance methods.

In [ ]:
class BankAccount:

  def __init__(self,account_holder_name,inital_balance):
    self.account_holder_name = account_holder_name
    self.inital_balance = inital_balance
    print(f"Thanks {account_holder_name}, your account has been created with a balance of {inital_balance}$ only. ")

  def deposit(self,amount):
    self.inital_balance += amount
    print("Amount Deposited Succesfully")
    print(f"New Balance: {self.inital_balance}$")

  def withdrawl (self,amount):
    if amount < self.inital_balance:
      self.inital_balance -= amount
      print("Amount Withdrawn Succesfully")
      print(f"New Balance: {self.inital_balance}$")
    else:
      print("insufficent balance! withdrawl failed. ")

  def balance_check (self):
    print(f"Current Balance: {self.inital_balance}$")

acc1 = BankAccount("Ali",100)
acc1.deposit(100)
acc1.withdrawl(100)
acc1.withdrawl(2000)
acc1.balance_check()


Thanks Ali, your account has been created with a balance of 100$ only. 
Amount Deposited Succesfully
New Balance: 200$
Amount Withdrawn Succesfully
New Balance: 100$
insufficent balance! withdrawl failed. 
Current Balance: 100$


102. Track Account Creation Using Class Variables
Enhance the BankAccount class by introducing a class-level attribute (shared across all instances) that keeps track of how many total accounts have been created. Each time a new account is instantiated, the counter should increase. Provide a method that returns the total number of accounts.

    Key Concepts: class variables vs instance variables, ClassName.variable, self.__class__.

In [ ]:
class BankAccount:

  total_accounts = 0

  def __init__(self,account_holder_name,inital_balance):
    self.account_holder_name = account_holder_name
    self.inital_balance = inital_balance
    print(f"Thanks {account_holder_name}, your account has been created with a balance of {inital_balance}$ only. ")
    BankAccount.total_accounts += 1

  def deposit(self,amount):
    self.inital_balance += amount
    print("Amount Deposited Succesfully")
    print(f"New Balance: {self.inital_balance}$")

  def withdrawl (self,amount):
    if amount < self.inital_balance:
      self.inital_balance -= amount
      print("Amount Withdrawn Succesfully")
      print(f"New Balance: {self.inital_balance}$")
    else:
      print("insufficent balance! withdrawl failed. ")

  def balance_check (self):
    print(f"Current Balance: {self.inital_balance}$")


  def get_accounts():
    print(f"Total Number for Accounts in {BankAccount.total_accounts}")

acc1 = BankAccount("Ali",100)
acc2 = BankAccount("Ali1",200)
acc3 = BankAccount("Ali2",300)
acc4 = BankAccount("Ali3",400)
BankAccount.get_accounts()

Thanks Ali, your account has been created with a balance of 100$ only. 
Thanks Ali1, your account has been created with a balance of 200$ only. 
Thanks Ali2, your account has been created with a balance of 300$ only. 
Thanks Ali3, your account has been created with a balance of 400$ only. 
Total Number for Accounts in 4


103. Human-Readable Account Display
Implement a method such that printing a BankAccount object gives a human-readable description like “Account holder: Alice, Balance: $2000” instead of a generic object reference. Also implement another method so the developer sees a more precise representation like “BankAccount('Alice', 2000)” for debugging.

    Key Concepts: Magic (dunder) methods __str__ and __repr__, string formatting, readability vs reproducibility.

In [ ]:
class BankAccount:

  total_accounts = 0

  def __init__(self,account_holder_name,inital_balance):
    self.account_holder_name = account_holder_name
    self.inital_balance = inital_balance
    print(f"Thanks {account_holder_name}, your account has been created with a balance of {inital_balance}$ only. ")
    BankAccount.total_accounts += 1

  def deposit(self,amount):
    self.inital_balance += amount
    print("Amount Deposited Succesfully")
    print(f"New Balance: {self.inital_balance}$")

  def withdrawl (self,amount):
    if amount < self.inital_balance:
      self.inital_balance -= amount
      print("Amount Withdrawn Succesfully")
      print(f"New Balance: {self.inital_balance}$")
    else:
      print("insufficent balance! withdrawl failed. ")

  def balance_check (self):
    print(f"Current Balance: {self.inital_balance}$")

  def __str__(self):
    return(f"Account holder: {self.account_holder_name}, Balance: ${self.inital_balance}")

  def __repr__(self):
    return(f"BankAccount('{self.account_holder_name}',{self.inital_balance})")

  @classmethod
  def get_accounts(cls):
    print(f"Total Number for Accounts in {cls.total_accounts}")

acc1 = BankAccount("Alice",2000)
print(str(acc1))
print(repr(acc1))

Thanks Alice, your account has been created with a balance of 2000$ only. 
Account holder: Alice, Balance: $2000
BankAccount('Alice',2000)


104. Combine Accounts with Operator Overloading
Make it so that adding two BankAccount instances with the + operator results in a new value representing their combined balances. For example, acc1 + acc2 should return the total balance of both accounts. You can later expand it to return a new BankAccount object as the sum.

    Key Concepts: __add__, operator overloading, intuitive interfaces.

In [ ]:
class BankAccount:

  total_accounts = 0

  def __init__(self,account_holder_name,inital_balance):
    self.account_holder_name = account_holder_name
    self.inital_balance = inital_balance
    print(f"Thanks {account_holder_name}, your account has been created with a balance of {inital_balance}$ only. ")
    BankAccount.total_accounts += 1

  def deposit(self,amount):
    self.inital_balance += amount
    print("Amount Deposited Succesfully")
    print(f"New Balance: {self.inital_balance}$")

  def withdrawl (self,amount):
    if amount < self.inital_balance:
      self.inital_balance -= amount
      print(f"{amount} Withdrawn Succesfully")
      print(f"New Balance: {self.inital_balance}$")
    else:
      print("insufficent balance! withdrawl failed.")

  def balance_check (self):
    print(f"Current Balance: {self.inital_balance}$")

  def __str__(self):
    return(f"Account holder: {self.account_holder_name}, Balance: ${self.inital_balance}")

  def __repr__(self):
    return(f"BankAccount('{self.account_holder_name}',{self.inital_balance})")

  def __add__(self,other):
    # return(self.inital_balance + other.inital_balance)
    name = self.account_holder_name + '_joined_' + other.account_holder_name
    balance = self.inital_balance + other.inital_balance
    return(self.__class__(name,balance))

  @classmethod
  def get_accounts(cls):
    print(f"Total Number for Accounts in {cls.total_accounts}")


acc1 = BankAccount("Alice",2000)
acc2 = BankAccount("Ali",2000)
# print(acc1+acc2) prints 4000

#update
acc3 = acc1 + acc2
print(acc3)



Thanks Alice, your account has been created with a balance of 2000$ only. 
Thanks Ali, your account has been created with a balance of 2000$ only. 
Thanks Alice_joined_Ali, your account has been created with a balance of 4000$ only. 
Account holder: Alice_joined_Ali, Balance: $4000


105.  Inheritance: SavingsAccount
Create a subclass SavingsAccount that inherits from BankAccount. Add a new attribute like interest_rate. Ensure you properly use super() in the constructor to initialize inherited attributes. Add a method to calculate interest after a period, e.g., calculate_interest(years).

    Key Concepts: Inheritance, constructor chaining using super(), extending functionality.

In [ ]:
class SavingsAccount (BankAccount):

  def __init__(self,account_holder_name,inital_balance,interest_rate):
    super().__init__(account_holder_name,inital_balance)
    self.interest_rate = interest_rate

  def calculate_interest(self, years):
    return self.inital_balance * years * self.interest_rate

acc1 = SavingsAccount("Alice",2000,0.33)
print(acc1.calculate_interest(6),"$ in total interest")

Thanks Alice, your account has been created with a balance of 2000$ only. 
3960.0 $ in total interest


106. Override Withdraw with Custom Rules
In SavingsAccount, override the withdraw() method such that if a user tries to withdraw more than a certain threshold (say, 80% of balance), a penalty is applied or the transaction is denied. This will show how you can change base class behavior in a subclass.

    Key Concepts: Method overriding, polymorphism, control flow in class methods.

In [ ]:
class SavingsAccount (BankAccount):

  def __init__(self,account_holder_name,inital_balance,interest_rate):
    super().__init__(account_holder_name,inital_balance)
    self.interest_rate = interest_rate

  def calculate_interest(self, years):
    return self.inital_balance * years * self.interest_rate

  def withdrawl(self, amount):
    if amount > 0.8 * self.inital_balance:
      penalty = 0.1 * self.inital_balance
      self.inital_balance -= penalty   # penalty
      super().withdrawl(amount)
      print(f"An Amount of {penalty}$ has been deducted")
    elif amount > self.inital_balance:
      print("Insufficent Balance! Withdrawl failed")
    else:
      super().withdrawl(amount)

acc1 = SavingsAccount("Alice1",200,0.33)
acc2 = SavingsAccount("Alice2",2000,0.33)
acc3 = SavingsAccount("Alice3",20000,0.33)
print()
acc1.withdrawl(101)
print()
acc2.withdrawl(1700)
print()
acc3.withdrawl(5000)


Thanks Alice1, your account has been created with a balance of 200$ only. 
Thanks Alice2, your account has been created with a balance of 2000$ only. 
Thanks Alice3, your account has been created with a balance of 20000$ only. 

101 Withdrawn Succesfully
New Balance: 99$


1700 Withdrawn Succesfully
New Balance: 100.0$

An Amount of 200.0$ has been deducted

5000 Withdrawn Succesfully
New Balance: 15000$



107. Validate Account Numbers with Static Method
Add a method that checks whether a provided bank account number is valid. For example, an account number must be exactly 10 digits. Since this check doesn’t depend on any instance or class state, define this using the @staticmethod decorator. Call it from outside the class without creating an object.

    Key Concepts: Static methods, utility functions, input validation, decorators.

In [ ]:
class SavingsAccount (BankAccount):

  def __init__(self,account_holder_name,inital_balance,interest_rate):
    super().__init__(account_holder_name,inital_balance)
    self.interest_rate = interest_rate

  def calculate_interest(self, years):
    return self.inital_balance * years * self.interest_rate

  def withdrawl(self, amount):
    if amount > 0.8 * self.inital_balance:
      penalty = 0.1 * self.inital_balance
      self.inital_balance -= penalty   # penalty
      super().withdrawl(amount)
      print(f"An Amount of {penalty}$ has been deducted")
    elif amount > self.inital_balance:
      print("Insufficent Balance! Withdrawl failed")
    else:
      super().withdrawl(amount)

  @staticmethod
  def account_num_validation(num):
    str_num = str(num)
    if len(str_num) == 10:
      return f"{num} is valid account number"
    else:
      return f"{num} is not valid account number"

print(SavingsAccount.account_num_validation(123456789))
print(SavingsAccount.account_num_validation(1234567890))

123456789 is not valid account number
1234567890 is valid account number


108.  Use Class Method to Access Class Variables
Now use a class method (marked with @classmethod) that returns the current total number of accounts (as added in Question 2). It should accept cls instead of self and interact with class-level data only.

    Key Concepts: @classmethod, cls, alternative constructors (bonus concept).

In [ ]:
class BankAccount:

  total_accounts = 0

  def __init__(self,account_holder_name,inital_balance):
    self.account_holder_name = account_holder_name
    self.inital_balance = inital_balance
    print(f"Thanks {account_holder_name}, your account has been created with a balance of {inital_balance}$ only. ")
    BankAccount.total_accounts += 1

  def deposit(self,amount):
    self.inital_balance += amount
    print("Amount Deposited Succesfully")
    print(f"New Balance: {self.inital_balance}$")

  def withdrawl (self,amount):
    if amount < self.inital_balance:
      self.inital_balance -= amount
      print(f"{amount} Withdrawn Succesfully")
      print(f"New Balance: {self.inital_balance}$")
    else:
      print("insufficent balance! withdrawl failed.")

  def balance_check (self):
    print(f"Current Balance: {self.inital_balance}$")

  def __str__(self):
    return(f"Account holder: {self.account_holder_name}, Balance: ${self.inital_balance}")

  def __repr__(self):
    return(f"BankAccount('{self.account_holder_name}',{self.inital_balance})")

  def __add__(self,other):
    # return(self.inital_balance + other.inital_balance)
    name = self.account_holder_name + '_joined_' + other.account_holder_name
    balance = self.inital_balance + other.inital_balance
    return(self.__class__(name,balance))

  @classmethod
  def get_accounts(cls):
    print(f"Total Number for Accounts in {cls.total_accounts} ") # {cls.balance_check()} would raise an error

acc1 = BankAccount("Alice",2000)
acc2 = BankAccount("Ali",2000)
ac1 =BankAccount("Alice",2000)
ac2 = BankAccount("Ali",2000)
BankAccount.get_accounts()

Thanks Alice, your account has been created with a balance of 2000$ only. 
Thanks Ali, your account has been created with a balance of 2000$ only. 
Thanks Alice, your account has been created with a balance of 2000$ only. 
Thanks Ali, your account has been created with a balance of 2000$ only. 
Total Number for Accounts in 4 


109. Abstraction with Abstract Base Class
Create an abstract class Vehicle using Python's abc module. This base class should declare two abstract methods: start() and stop(). Then create subclasses Car and Bike, each implementing those methods in their own way. Attempting to instantiate Vehicle directly should raise an error.

    Key Concepts: Abstract classes, ABC, @abstractmethod, forcing subclasses to implement methods.

In [ ]:
# havent done this yet, and since it is not a core concept and is done via a module, im not doing ity yet.

110. Multiple Inheritance and Method Resolution
Design two independent classes: GPS with methods like navigate(), and MusicSystem with methods like play_music(). Then create a SmartCar class that inherits from both. Call both methods from the SmartCar instance. Also, use super() in all constructors and observe Python's method resolution order (MRO) to understand how it handles such hierarchies.

    Key Concepts: Multiple inheritance, MRO, super(), diamond problem (brief intro)

In [ ]:
class GPS:

  def __init__(self):
    print("GPS intialized")
    super().__init__()

  def navigate(self):
    print("Navigation Started")

class MusicSystem:

  def __init__(self):
    print("Music intialized")
    super().__init__()

  def play_music(self):
    print("Music Started")

class SmartCar(GPS,MusicSystem):

  def __init__(self):
    print("Car Ready")
    super().__init__()

yesla = SmartCar()
print(SmartCar.mro())

Car Ready
GPS intialized
Music intialized
[<class '__main__.SmartCar'>, <class '__main__.GPS'>, <class '__main__.MusicSystem'>, <class 'object'>]


## File Handling

File Handling Questions (111–115)





111. Write and Read File (Warm-Up)



          Objective: Practice open(), file modes 'w' and 'r', write(), read(), and close().
          Task:
          Open greetings.txt in write mode.
          Write the string "Hello, Shah G!".
          Close the file.
          Re-open in read mode and print its contents.
          
          Goal: Understand the full flow of writing → closing → reading.

In [21]:
from os import close
f = open("source.txt" , 'w')
f.write("Hello. Shah G!")
f.close()

f = open("source.txt" , 'r')
print(f.read())
f.close()
f.closed

# for next task
f = open("source.txt", 'w')
for i in range(10):
  f.write("\nHello. Shah G!")
f.close()
f.closed


Hello. Shah G!


True

 112. Copy File Content Line by Line
          Objective: Practice readline() or readlines() and writing to another file using 'wt'.
          Task:
          Read from a file named source.txt line by line.
          Write each line into backup.txt.
          
          Goal: Simulate file backup by line, and strengthen looping through files.

In [22]:
data = []
with open("source.txt", 'r') as f:
  data = f.readlines()

with open("backup.txt", 'wt') as f:
  for i in data:
    f.write(f"{i}")

with open("backup.txt", 'r') as f:
  print(f.read())


Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!
Hello. Shah G!


113. Count Words in a File
          Objective: Practice reading data and using split() and loops.
         
          Task:
          Open any .txt file.
          Count total words using len(f.read().split()).
          
          Goal: Combine file I/O with string methods and processing.

In [24]:
with open("backup.txt", 'r') as f:
  print(len(f.read().split()))

30


114. Append User Input to Log File
          Objective: Practice file mode 'a', taking user input, and writing
          with newline characters.
          Task:
          Ask user to enter a message.
          Append it to log.txt.
          Rerun the program to test whether new lines get added without erasing
          previous ones.
         
          Goal: Understand file persistence and appending patterns.

In [26]:
input_1 =  input("enter data for log.txt: ")

with open("log.txt", 'a') as f:
  f.write(f"{input_1}\n")

input_1 =  input("enter data for log.txt again: ")

with open("log.txt", 'a') as f:
  f.write(f"{input_1}\n")

with open("log.txt" , 'r') as f:
  print(f.read())



enter data for log.txt: this is 1st line
enter data for log.txt again: this is 2nd line
this is first input
this is second input
this is 1st line
this is 2nd line



 115. File Pointer Positioning with tell() and seek()
          Objective: Deep dive into file cursor control.
         
          Task:
          Open a file for reading.
          Use .tell() to see pointer position.
          Read 5 chars.
          Use .tell() again.
          Use .seek(0) and read entire file again.
         
          Goal: See how file pointer shifts and how to reset it.

In [27]:
with open("log.txt", 'r') as f:
  print(f.tell())
  print(f.read(5))
  print(f.tell())
  print(f.seek(0))
  print(f.read())

0
this 
5
0
this is first input
this is second input
this is 1st line
this is 2nd line



## Exception Handling Questions (116–120)

File Handling Questions (116–120)



116. Catch Division by Zero Error
          Objective: Understand try-except with a specific error class (ZeroDivisionError).
          Task:

          Write a function that divides two numbers.

          Wrap the division in a try block.

          Catch zero division and print "Cannot divide by zero" if error occurs.
          Goal: Prevent program crash and control error output.


In [30]:
def div(a,b):
  try:
    return a/b
  except Exception as e:
    return f"Result for {b}/{a} :{b/a} ----> {a}/{b} caused an error ---> {e}"

print(div(9,0))
print(div(9,3))



Result for 0/9 :0.0 ----> 9/0 caused an error ---> division by zero
3.0


 117. Handle Multiple Exceptions (Input Errors)
          Objective: Handle user input-related exceptions cleanly.
          Task:

          Ask user to input a number.

          Use int(input(...)).

          Catch ValueError (if input is not a number).

          Catch KeyboardInterrupt (if user hits Ctrl+C).
          Goal: Show different messages for different error types. Real-world input validation.


In [37]:
import sys
try:
  check = int(input("Enter a number: "))
except ValueError as e:
  print("Error, you didnt enter a number: " , e)
except KeyboardInterrupt as e:
  print("Error you stopped the program form executing " , e)

Error you stopped the program form executing  Interrupted by user
Enter a number: fuck


118. File Not Found Exception
          Objective: Handle file-related errors (FileNotFoundError).
          Task:

          Try opening data.txt in read mode.

          If not found, catch the exception and print a user-friendly message.
          Bonus: Ask user to retry entering filename.
          Goal: Gracefully handle missing resources.



In [38]:
try:
  with open("data.txt" ,'r') as f:
    f.read()
except FileNotFoundError as e:
  print(f"File doesnt exist --- > {e}")


File doesnt exist --- > [Errno 2] No such file or directory: 'data.txt'


 119. Demonstrate Finally Block
            Objective: Understand the unconditional execution of finally for cleanups.
            Task:

            Open a file (even a non-existent one).

            Catch any error.

            In the finally block, print "Operation Finished" or close file if it was opened.
            Goal: Practice cleanup logic — common in file/database tasks.

In [39]:
try:
  f = open("data.txt" , 'w')
  print(f.read())
except Exception as e:
  print("Error Ocurred" , e)
finally:
  print("Opeartion completed!")
  f.close()

Error Ocurred not readable
Opeartion completed!


120. Print Custom Exception Message
          Objective: Practice Exception as e pattern for debugging.
          Task:

          Cause a TypeError by adding an integer to a string.

          Use try-except to catch it.

          Print: "The error occurred is: <error>"
          Goal: Learn how to fetch error messages dynamically and display them.

In [40]:
try:
  20 + "20"
except Exception as e:
  print(f"The error occurred is: {e}")

The error occurred is: unsupported operand type(s) for +: 'int' and 'str'
